### PR TITLE
net/sock/dtls: allow using multiple credentials

### PR DESCRIPTION
**Pull request by leandrolanzieri**
### Contribution description
Currently only one credential can be associated to each DTLS sock. This limits certain use cases that need to establish multiple connections on a single sock using potentially multiple credentials.

This PR proposes an extension of the sock DTLS API to allow registering multiple credentials to be used on a single sock. An application may register callbacks to specify a certain credential from the list based on the session information (and, when using PSK, based on the hint that may have been sent).

Additionally, now a server can send a PSK Identity hint to a client during the handshake, to help deciding on the PSK Identity that should be used.

Finally, the example application has been extended to show these functionalities.

### Testing procedure
- The example application should be working both when using PSK and ECC (for more information enable the debug messages in the sock_dtls.c file).
- Check that the correct hint is being sent in PSK
- Check that the correct RPK is being used in ECC
- Read the documentation
- A single credential use case should still work seamlessly

### Issues/PRs references
None so far

### CODE REVIEW

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [inline pkg/tinydtls/contrib/sock_dtls.c:220]: Note to self: prioritize credential by hint.

**Comment by leandrolanzieri** [comment]
@janosbrodbeck the new definitions are not split into their own header, this should avoid the issues when using Sock Async.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (pokgak) [inline pkg/tinydtls/contrib/sock_dtls.c:406]: I'm thinking if we need to call `sock_dtls_add_credential` for each new `credman_add` anyways, would it be better to include `credman_add` in `sock_dtls_add_credential` also? The parameter list might need to change then though.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (pokgak) [inline pkg/tinydtls/contrib/sock_dtls.c:415]:
```suggestion
    DEBUG("sock_dtls: could not add new credential\n");
```
minor

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (pokgak) [inline pkg/tinydtls/contrib/sock_dtls.c:83]: If the corresponding credentials are removed with `sock_dtls_remove_credential`, will it also be removed from `_ecdsa_keys`?

**Review by pokgak — CHANGES_REQUESTED** [review]
Looks good overall.

**Review by janosbrodbeck — CHANGES_REQUESTED** [review]
`tests/pkg_tinydtls_sock_async/dtls-server.c` needs adjustments, because the PSK matching changed. The server in the test does not provide an identity for the credential. Don't know why the ID is missing but now it leads with the current changes to not being able to find a matching credential, since it compares with an empty identity. I suspect that since only one credential was possible before, the only existing key was simply used without identity comparison.

```
static const uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;
```
needs to be added to `pkg_tinydtls_sock_async/dtls-server.c`. And the ID then needs to be added to the credential.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (janosbrodbeck) [reply pkg/tinydtls/contrib/sock_dtls.c:406]: It is completely legitimate to use credentials for multiple sockets. Running `credman_add()` every time would mean that every time you reuse a credential, you would first have to find out that the credential already exists.

**Comment by janosbrodbeck** [comment]
Looks good to me. Existing code is still working (except the test, but I'll leave that out) and the API is easy to use. Already have a WIP build for my gcoap-dtls running which supports the new functionality.

However, I am for a small extension in `sock_dtls_create()` as I found in a small issue:

When creating a new DTLS socket it is mandatory to provide a credman tag, it includes `sock_dtls_add_credential()`. That was okay when there could only be one credential per socket. For gcoap I defined a pre-defined tag and the user could add their credential and tag it with the gcoap tag. But now there is no need for that anymore, I would prefer to have the possibility to create a socket with an empty tag list. The user has to create the credential anyway, and I see no reason to make the usage of a pre-defined tag mandatory. I would like to leave the tag creation up to the user.

But the current API does not provide that functionality. For gcoap I would have to create the socket with an dummy tag and then remove it with `sock_dtls_remove_credential()` to get an empty tag list. I would be in favour of changing the API tbh.

We should add an option to allow socket creation with empty credential list, e.g. define a `CREDMAN_NO_TAG` and parse this define in the `sock_dtls_create()` to "no tag".

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:83]: Currently not. It should actually not interfere with an ongoing session. Now that you mention this, probably `sock_dtls_session_destroy` should mark as free the previously used structure. Does tinyDTLS expose a pointer to the provided structure? If that's not the case, then probably that state should be kept externally (session <-> ecdsa struct).

**Comment by leandrolanzieri** [comment]
> `static const uint8_t psk_id_0[] = PSK_DEFAULT_IDENTITY;` needs to be added to `pkg_tinydtls_sock_async/dtls-server.c`. And the ID then needs to be added to the credential.

Added in 26cf8cf9f505576a6b0b073a423a601c996e186a.

> But the current API does not provide that functionality. It should be fairly easy to achieve, since `CREDMAN_TAG_EMPTY` already exists. We should parse this option in `sock_dtls_create()` to allow the creation of a socket with an empty credential list. And should also add this possibility to the documentation.

Added in b051dae589485da83e41c1283d5908399f5db8c3.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: Done in 01dc4dfd5d454871f054d7188fd24605f01925e2

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: In d054da2597a45cb3acc4610d80a6df80293738a7 I modified the logic to keep a pointer to the session during the handshake. I did not find any documentation regarding when the key can be freed, currently this is done when the handshake is complete. Maybe we should check what happens on reconnections.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (janosbrodbeck) [inline pkg/tinydtls/contrib/sock_dtls.c:153]: `_ecdsa_keys` is not declared when `CONFIG_DTLS_ECC` is not set, because `_ecdsa_keys` is declared in a `#ifdef` block. This leads to compilation errors currently.

**Review by janosbrodbeck — CHANGES_REQUESTED** [review]

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:153]: Yep, I'll change it to `ifdef`

Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:154]: Done

File: sys/include/net/sock/dtls.h
Comment (pokgak) [inline sys/include/net/sock/dtls.h:484]: This might need updating after the changes to allow `CREDMAN_TAG_EMPTY` on `sock_dtls_create`.

File: sys/include/net/sock/dtls.h
Comment (pokgak) [inline sys/include/net/sock/dtls.h:495]: ditto

**Review by pokgak — CHANGES_REQUESTED** [review]
Almost done. Just need to update the docs.

**Review by janosbrodbeck — APPROVED** [review]
Besides the documentation updates that pokgak suggested, I am very happy.

Tested it again with `examples/dtls-sock`, `tests/pkg_tinydtls_sock_async` and my gcoap-dtls PR without adaption changes and with some first integration changes. The right credentials are used, and existing single credential use cases are also working without changes.

File: sys/include/net/sock/dtls.h
Comment (leandrolanzieri) [reply sys/include/net/sock/dtls.h]: Not sure what exactly needs to change here. I made some adaptions to the doc in 74ea5086deba55b80c490b456d0d84b9f229f5ff. Could you take a look?

**Comment by leandrolanzieri** [comment]
I added another function to get a read-only array of the registered credentials on a sock.

**Review by pokgak — APPROVED** [review]
You can squash now.

**Comment by leandrolanzieri** [comment]
Any maintainer that would like to take a look? @miri64? @cgundogan?

File: pkg/tinydtls/include/sock_dtls_types.h
Comment (miri64) [inline pkg/tinydtls/include/sock_dtls_types.h:80]: I know that this would probably mean a redesign of the `credman` API so I actually would prefer that to be a follow-up, if that is the case, but: Would it make sense to have `tags` (and also maybe `psk_hint`) be a linked list instead of an array?

Comment (leandrolanzieri) [reply pkg/tinydtls/include/sock_dtls_types.h:80]: > I know that this would probably mean a redesign of the `credman` API so I actually would prefer that to be a follow-up, if that is the case, but: Would it make sense to have `tags` (and also maybe `psk_hint`) be a linked list instead of an array?

Not sure if that would actually affect `credman`'s API. The `psk_hint` is just a string, so it should stay as is. It seems a bit more cumbersome from the user perspective having to allocate each list element to register a new tag.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (PeterKietzmann) [inline pkg/tinydtls/contrib/sock_dtls.c:349]: Please document this. You assign the key if there is a free session "slot"?

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (PeterKietzmann) [inline pkg/tinydtls/contrib/sock_dtls.c:479]: Really have to copy the tags?

File: sys/include/net/sock/dtls.h
Comment (PeterKietzmann) [inline sys/include/net/sock/dtls.h:46]: Change order of first and second bullet to match the above server operations, or does this break the client logic?

File: sys/include/net/sock/dtls.h
Comment (PeterKietzmann) [inline sys/include/net/sock/dtls.h:490]: Reference to the RFC/Section?

File: sys/include/net/sock/dtls.h
Comment (PeterKietzmann) [inline sys/include/net/sock/dtls.h:493]: I would change the order of arguments. The client can register the callback and on execution, it can utilize the hint

**Review by PeterKietzmann — COMMENTED** [review]
@leandrolanzieri nice contribution! I've left a first round of comments but didn't test it yet. Overall the code looks very good

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:479]: How to store them if not?

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:245]: Hmm I think there is a missing break after line 237. If there is no hint provided, and a valid PSK credential is found, then we should look no further.

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:349]: Yes, tinyDTLS requires that the structure holding the key is allocated by the callback, so I introduced a pool of them.

File: sys/include/net/sock/dtls.h
Comment (leandrolanzieri) [reply sys/include/net/sock/dtls.h:46]: Isn't it matching the server order? 

0. set hint (not available in client)
1. add extra credentials
2. set PSK callback (not available in server)
3. set RPK callback

File: sys/include/net/sock/dtls.h
Comment (leandrolanzieri) [reply sys/include/net/sock/dtls.h:493]: Sorry, I don't understand, can you suggest the change you imagine here?

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:352]: Comment added

File: sys/include/net/sock/dtls.h
Comment (leandrolanzieri) [reply sys/include/net/sock/dtls.h:490]: Reference added

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c]: Changed

File: pkg/tinydtls/contrib/sock_dtls.c
Comment (leandrolanzieri) [reply pkg/tinydtls/contrib/sock_dtls.c:246]: Fixed

File: sys/include/net/sock/dtls.h
Comment (PeterKietzmann) [reply sys/include/net/sock/dtls.h:494]: What I meant was: The client application does not decide which credential to use by **registering** the callback but by **executing** the callback that has been registered **beforehand**. Never mind, reading this again I think it is still understandable this way and shorter.

**Comment by PeterKietzmann** [comment]
@leandrolanzieri thanks again for this nice contribution! I've tested multiple scenarios on native and iotlab-m3, all seem to work fine (see report below). The example README could get some enhancement, though. What to expect from a matching (or non-matching) hint, where to change it for experimentation, how to enable to elliptic curve credentials...

<details>
  <summary>examples/dtls-sock, native, PSK</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
Client got hint: hint
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: comparing to tag 3, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"
Terminating

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: error receiving UDP packet: -110
sock_dtls: psk hint request
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
New client connected
sock_dtls: decrypted message arrived
Received 3 bytes -- (echo)
```

dtls


</details>

<details>
  <summary>examples/dtls-sock, native, PSK (w/o hint)</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: found

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: error receiving UDP packet: -110
sock_dtls: psk hint request
sock_dtls: no hint provided
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: invalid credential params for type 2
```
dtls_nohint


</details>


<details>
  <summary>examples/dtls-sock, native, ECC</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: event connect
sock_dtls: get ECDSA key
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"

```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::98dd:b3ff:fec3:e399%5]:20220
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Wrong_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: found

```
dtls_ecc

</details>

<details>
  <summary>examples/dtls-sock, iotlab-m3, PSK</summary>
  
  ## Client

```
> main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::ec1a:a605:46d1:c2d9 "ABC"
dtlsc fe80::ec1a:a605:46d1:c2d9 "ABC"
sock_dtls: credential added in position 1
sock_dtls: event connect
sock_dtls: psk id request
sock_dtls: requesting the application
From [fe80::ec1a:a605:46d1:c2d9%5]:20220
Client got hint: hint
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Wrong_identity
sock_dtls: comparing to tag 3, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Connection to server successful
Sent DTLS message
sock_dtls: decrypted message arrived
Received 3 bytes: "ABC"
Terminating
```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  5  HWaddr: 42:D9  Channel: 26  Page: 0  NID: 0x23  PHY: O-QPSK

          Long HWaddr: EE:1A:A6:05:46:D1:C2:D9
           TX-Power: 0dBm  State: IDLE  max. Retrans.: 3  CSMA Retries: 4
          AUTOACK  ACK_REQ  CSMA  L2-PDU:102  MTU:1280  HL:64  6LO
          IPHC
          Source address length: 8
          Link type: wireless
          inet6 addr: fe80::ec1a:a605:46d1:c2d9  scope: link  VAL
          inet6 group: ff02::1

> sock_dtls: psk hint request
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
New client connected
sock_dtls: decrypted message arrived
Received 3 bytes -- (echo)
sock_dtls: error receiving UDP packet: -116
```

</details>

<details>
  <summary>tests/pkg_tinydtls_sock_async, native, PSK</summary>
  
  ## Client

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlsc fe80::98dd:b3ff:fec3:e399
dtlsc fe80::98dd:b3ff:fec3:e399
usage dtlsc <addr> <data>
> dtlsc fe80::98dd:b3ff:fec3:e399 ""^[[DABC
dtlsc fe80::98dd:b3ff:fec3:e399 "ABC
sock_dtls: event connect
> sock_dtls: psk id request
sock_dtls: trying to get first PSK credential
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 2, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Session handshake received
Connection to server successful
Sending data ABC"
Sent DTLS message
Session became ready: [fe80::98dd:b3ff:fec3:e399%7]:20220
DTLS message was sent
sock_dtls: decrypted message arrived
Received 6 bytes: ABC"
Terminating session
```

## Server

```
main(): This is RIOT! (Version: 2021.01-devel-3036-gf749dc-HEAD)
DTLS sock example application
All up, running the shell now
> dtlss start
dtlss start
> ifconfig
ifconfig
Iface  7  HWaddr: 9A:DD:B3:C3:E3:99 
          L2-PDU:1500  MTU:1500  HL:64  Source address length: 6
          Link type: wired
          inet6 addr: fe80::98dd:b3ff:fec3:e399  scope: link  VAL
          inet6 group: ff02::1
          inet6 group: ff02::1:ffc3:e399
          
> sock_dtls: psk hint request
sock_dtls: no hint provided
sock_dtls: psk key request
sock_dtls: looking for key for ID: Client_identity
sock_dtls: comparing to tag 10, with ID: Client_identity
sock_dtls: found
sock_dtls: event connected
Session handshake received
New client connected
Session became ready: [fe80::d0f6:c9ff:fe2a:984a%7]:12345
sock_dtls: decrypted message arrived
Received 6 bytes -- (echo)
Session was destroyed by peer: [fe80::d0f6:c9ff:fe2a:984a%7]:12345
```

</details>

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri thanks again for this nice contribution! I've tested multiple scenarios on native and iotlab-m3, all seem to work fine (see report below). The example README could get some enhancement, though. What to expect from a matching (or non-matching) hint, where to change it for experimentation, how to enable to elliptic curve credentials...

@PeterKietzmann thanks for the thorough testing! In 4a153eaec038ef040642bf64d0031f2254034bed I added a couple of sections in the application's README. Could you give it a look?

**Comment by PeterKietzmann** [comment]
@leandrolanzieri thanks, looks better! Maybe introduce PSK and RPK as abbreviations. Mention that the hint (string) could also be changed, you could also explain that it is printed on the client (for verification by the user). I leave this up to you... Please squash directly

**Comment by leandrolanzieri** [comment]
> @leandrolanzieri thanks, looks better! Maybe introduce PSK and RPK as abbreviations. Mention that the hint (string) could also be changed, you could also explain that it is printed on the client (for verification by the user). I leave this up to you... Please squash directly

Added the suggestions and squashed directly!

File: sys/include/net/sock/dtls/creds.h
Comment (PeterKietzmann) [inline sys/include/net/sock/dtls/creds.h:91]: This is a function for the server only, right? Would it be reasonable to add this to the function name? Like you did for the client functions (for example sock_dtls_set_client_psk_cb)

**Review by PeterKietzmann — APPROVED** [review]
ACK from my side. Anyone else who wants to give this PR a look before merge?

File: sys/include/net/sock/dtls/creds.h
Comment (leandrolanzieri) [reply sys/include/net/sock/dtls/creds.h]: I renamed it to `sock_dtls_set_server_psk_id_hint`

**Comment by PeterKietzmann** [comment]
Well then, go!

**Pull request merged**
